tests/test_io.py

```diff
@@ -0,0 +1,63 @@
+from typing import Callable, Dict, List
+
+import pytest
+import ulist as ul
+from ulist.utils import check_test_result
+
+
+@pytest.mark.parametrize(
+    "test_method, args, kwargs, expected_value",
+    [
+        (ul.read_csv, (), {
```

Comment:
I am wondering if the error messages would  be raised as expected, if the `schema` dtype doe not match actual dtype. So shall we add some more tests like [this](https://github.com/Rust-Data-Science/ulist/blob/ab57251136d375a5e47a61af9a3262394795c0db/tests/test_exceptions.py#L86)?

Reply:
Other cases could be:

- The `schema.len()` is greater than the number of columns in the `.csv` file;
- The `schema.len()` is less than the number of columns in the `.csv` file;
- The column name does not match.

Reply:
> Other cases could be:
> 
>     * The `schema.len()` is greater than the number of columns in the `.csv` file;
> 
>     * The `schema.len()` is less than the number of columns in the `.csv` file;
> 
>     * The column name does not match.

Hello,
the strategy I chose is:

1. If the field in `schema` doesn't exists in the `.csv` file, then `ulist` will return an empty list for it, such as `{..., "bar": [], ...}`
2. If the field in `.csv` file doesn't exists in `schema`, then `ulist` will ignore it

So currently, cases above can not cause an exception.

Reply:
For 2., I agree that the result of `read_csv` could be a subset of then content of `.csv` file, so could you also add another test case, where the schema has fewer columns than the csv file?

Reply:
For 1., let's do some research on mainstream data analysis, data processing libs or even databases to see what kind of behavior would be better.

1. Pandas, the most popular `DataFrame` lib: The `read_csv` function will always read **all the columns** from csv file. No matter how many col names are in the schema. See the [Doc](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html)
Suppose the `tmp.csv` file contains only columns `a` and `b`
```Python3
>>> import pandas as pd
>>> pd.read_csv("tmp.csv")
   a   b
0  1   2
>>> pd.read_csv("tmp.csv", dtype={'a': int, 'b': int})
   a   b
0  1   2
>>> pd.read_csv("tmp.csv", dtype={'a': int, 'b': int, 'c': int})
   a   b
0  1   2
>>> pd.read_csv("tmp.csv", dtype={'a': int, 'b': int, 'c': int, 'd': int})
   a   b
0  1   2
>>> pd.read_csv("tmp.csv", dtype={'c': int, 'd': int})
   a   b
0  1   2
```

2. SQL, the most used language for database: It does a very strict check for column names, and will raise an error if the name does not exist in the data base. For example, if there is table named `tmp` which contains column `a` and `b`.  The following script will just not work.
```sql
SELECT CAST(tmp.c AS INT) as c  from tmp  
```
If you don't mind, would you help test the behavior of [pyarrow](https://arrow.apache.org/docs/python/generated/pyarrow.csv.read_csv.html), which is also one of the best data processing lib.

Thanks so much~

Reply:
And please feel free to investigate any other popular projects and share your idea here. The reason why we have to be cautious is because in real industry usage, the number of columns could be 10, 20 , 100 or even thousands, and it's very common that the developers types the incorrect col names in the `schema`. So simply ignore the typo, raise a warning or an error is an important topic to discuss.

Reply:
> For 2., I agree that the result of `read_csv` could be a subset of then content of `.csv` file, so could you also add another test case, where the schema has fewer columns than the csv file?

Sure, the test case is appended in the newest commit b86fc43: 
```python
...
            (ul.read_csv, (), {  # schema.len() < field.len()
                "path": "./test_csv/04_test_nan.csv",
                "schema": {"int": "int",
                           "bool": "bool"}
            }, {
                "int": [None, 2, 3, 4],
                "bool": [True, False, True, None]
            }),
...
```

Reply:
> For 1., let's do some research on mainstream data analysis, data processing libs or even databases to see what kind of behavior would be better.
> 
>     1. Pandas, the most popular `DataFrame` lib: The `read_csv` function will always read **all the columns** from csv file. No matter how many col names are in the schema. See the [Doc](https://pandas.pydata.org/docs/reference/api/pandas.read_csv.html)
>        Suppose the `tmp.csv` file contains only columns `a` and `b`
> 
> 
> ```python
> >>> import pandas as pd
> >>> pd.read_csv("tmp.csv")
>    a   b
> 0  1   2
> >>> pd.read_csv("tmp.csv", dtype={'a': int, 'b': int})
>    a   b
> 0  1   2
> >>> pd.read_csv("tmp.csv", dtype={'a': int, 'b': int, 'c': int})
>    a   b
> 0  1   2
> >>> pd.read_csv("tmp.csv", dtype={'a': int, 'b': int, 'c': int, 'd': int})
>    a   b
> 0  1   2
> >>> pd.read_csv("tmp.csv", dtype={'c': int, 'd': int})
>    a   b
> 0  1   2
> ```
> 
>     2. SQL, the most used language for database: It does a very strict check for column names, and will raise an error if the name does not exist in the data base. For example, if there is table named `tmp` which contains column `a` and `b`.  The following script will just not work.
> 
> 
> ```sql
> SELECT CAST(tmp.c AS INT) as c  from tmp  
> ```
> 
> If you don't mind, would you help test the behavior of [pyarrow](https://arrow.apache.org/docs/python/generated/pyarrow.csv.read_csv.html), which is also one of the best data processing lib.
> 
> Thanks so much~

^_^ I am glad to do this, but may be a little latter. I never think about this before, honestly. I am so inspired by your scientific attitude